File: Easy/206. Reverse Linked List/solution.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    
    public ListNode ReverseListHelper(ListNode preNode, ListNode currentNode){
        if (currentNode == null){
            return preNode;
        }
        ListNode nextNode = currentNode.next;
        currentNode.next = preNode;
        return ReverseListHelper(currentNode, nextNode);
    }
    
    public ListNode ReverseList(ListNode head) {
        ListNode preNode = null;
        return ReverseListHelper(preNode, head);
    }
}

File: Easy/206. Reverse Linked List/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseListHelper(ListNode* preNode, ListNode* curNode){
        if (curNode == NULL)    return preNode;
        ListNode* nextNode = curNode -> next;
        curNode -> next = preNode;
        
        return reverseListHelper(curNode, nextNode);
    }
    
    ListNode* reverseList(ListNode* head) {
        return reverseListHelper(NULL, head);
    }
};